### ISSUE TITLE
there is one dart process use 100% cpu after upgrade to v3.12.2 in mac

### ISSUE THREAD

**Issue by wanjm** [Dart-Code/Dart-Code#2624]
there is one dart process use 100% cpu after upgrade to v3.12.2 in mac;

and functions can't jump to its definition any more;

**Comment by DanTup** [comment]
Can you confirm what platform you're on, and what the full command line of the Dart process using the CPU is (eg. `ps -x | grep dart` or on Windows checking the Command Line column in Task Manager)?

**Comment by wanjm** [comment]
501 32038 31605   0 Fri11AM ??       471:06.68 /opt/google/flutter/bin/cache/dart-sdk/bin/dart /opt/google/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot --client-id=VS-Code --client-version=3.12.2


which use 100% cpu;

**Comment by wanjm** [comment]
one workspace with more project may trigger such problem ; 
I have a workspace with five project, only one is flutter(dart). then dart use 100%cpu; then i remove one project from the workspace, kill the dart, and restart it works ok;

**Comment by wanjm** [comment]
dart begins to run though there is no dart project in current workspace,  and use 100% cpu;

**Comment by DanTup** [comment]
Could you try enabling logging with the `dart.analyzerLogFile` setting (see https://dartcode.org/docs/logging/#analyzer) and provide a log file when it crashes?

Note: The log file may contain parts of your source code, so it's best to avoid private projects when reproducing this.

**Comment by wanjm** [comment]
in vscode , and flutter project folder to workspace, and then add one js project to project, then dart will use 100% cpu.

**Comment by DanTup** [comment]
@wanjm is it possible you captured the general log (https://dartcode.org/docs/logging/#extension) instead of the analyzer log (https://dartcode.org/docs/logging/#analyzer)? The log you sent doesn't seem to include the analyzer traffic.

Also - can you confirm whether there's anything special about the JS project? Is it on a local disk or a remote disk? Might it contain any symbol links?

**Comment by wanjm** [comment]
emailed the log.
the bug happens on both mac and windown;

**Comment by DanTup** [comment]
I can't see anything unusual in the log. Can you confirm exactly what steps you're performing, including which folder paths you're initially opening and then adding that triggers the issue?

It's a lot more verbose, but the [instrumentation log](https://dartcode.org/docs/logging/#analyzer-instrumentation) might help. In the past I've seen issues where the file system had symlinks that loop back to parent folders (thus creating infinite loops when walking the folder) that have caused issues like this. The last few lines of the instrumentation log might help indicate if that's happening.

You could also try opening only the JS project, but add a single `.dart` files to it - that will trigger the extension to load, so you can see if it happens with just that project on its own (this should produce a smaller log and be easier to debug with, if so).

Also - can you confirm if it happens only with that specific JS project, or any other non-Dart folder you add?

Thanks!

**Comment by wanjm** [comment]
when I just open one of my js project, there is no dart in it, then cpu will be high;

then I removed the node_modules directory and open again everything is ok;

I reinstall the node_modules using " npm install" and open the project again, everything is ok;

I removed the node_modules directory and reintall the node_modules using "cnpm install"  (which is previde by alibaba") , and reopen the directory, cpu usage be high;

I create on simple project  with package.json then it is reproduced; and the project has email to you ;

https://github.com/cnpm/cnpm, it is the command i used;

**Comment by DanTup** [comment]
I didn't get the sample project, but could you try running `find . -type l -ls` inside the project to see if there are any symlinks? If so, I may be able to reproduce it by setting up my own links. I'm not sure if the issue is just that `node_modules` is huge, or whether it has cycles. The fix for each may be different.

Thanks!

**Comment by wanjm** [comment]
use cnpm to install node_modules, all are soft link;
![image](https://user-images.githubusercontent.com/3118573/93159689-5b01c900-f741-11ea-98f4-407994480f9b.png)

**Comment by wanjm** [comment]
this is what I am using;

"dependencies": {
    "adm-zip": "^0.4.7",
    "ali-oss": "^6.1.1",
    "amqplib": "^0.5.5",
    "body-parser": "^1.15.1",
    "canvas": "^2.6.0",
    "co": "^4.6.0",
    "commander": "^4.0.1",
    "crc-32": "^1.2.0",
    "cuint": "^0.2.2",
    "express": "^4.13.4",
    "gif-frames": "^1.0.1",
    "image-size": "^0.7.4",
    "mkdirp": "^0.5.1",
    "modesl": "^1.1.8",
    "moment": "^2.13.0",
    "multer": "^1.1.0",
    "objectid": "^3.2.1",
    "prpc": "0.1.30",
    "request-promise": "^4.2.5",
    "send": "^0.17.1",
    "tls-sig-api-v2": "^1.0.0",
    "ua-parser-js": "^0.7.17",
    "underscore": "^1.8.3",
    "uuid": "^3.1.0",
    "ws": "^7.0.0"
  },
  "devDependencies": {
    "@types/eslint": "^7.2.0",
    "@types/node-ssh": "^7.0.0",
    "@types/request-promise": "^4.1.46",
    "compressing": "^1.5.1",
    "eslint": "^7.3.1",
    "eslint-config-standard": "^14.1.1",
    "eslint-plugin-import": "^2.22.0",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-promise": "^4.2.1",
    "eslint-plugin-standard": "^4.0.1",
    "mocha": "^7.0.1",
    "node-ssh": "^10.0.2",
    "pre-commit": "^1.2.2"
  },

**Comment by DanTup** [comment]
I tried using cnpm to install those modules in a sample app, but I don't seem to have the same symlinks as you (nor do do I get the high CPU issue).

Are you able to share a sample project as a zip file that reproduces this for you that I can just extract, run `cnpm install` and hopefully have the exact same setup as you?

Can you also confirm exactly what folder(s) you're opening in VS Code - is it just this one, or along with another Dart project? Can you reproduce it by *only* opening the JS project (you can add a .dart file to it to force the extension to activate)? Thanks!

**Comment by wanjm** [comment]
[tempjs.tar.gz](https://github.com/Dart-Code/Dart-Code/files/5237087/tempjs.tar.gz)

**Comment by wanjm** [comment]
only open this js project will trigget the issue;
I think it is softlink caused the problem.

**Comment by DanTup** [comment]
Thanks - I was able to repro now - though I had to add a `.dart` file to the folder to force the extension to activate. I did find some circular links, for example:

tempjs/node_modules/_es-abstract@1.17.6@es-abstract/node_modules/string.prototype.trimend/node_modules/es-abstract

The `es-abstract` folder at the end if a link back to `_es-abstract@1.17.6@es-abstract`, so the server is following these links and getting stuck in an infinite loop.

I'll close this as a dupe of https://github.com/dart-lang/sdk/issues/42079 and we can discuss further there.

Thanks!
